### ISSUE TITLE
WebDriver:NewSession fails on unrooted Android devices ("this.curBrowser is null")

### ISSUE THREAD

**Issue by sparsh00** [mozilla/geckodriver#1858]
## System

* Version:  0.27.0 
* Platform:  macos 10.13.6
* Firefox:  86.1.1
* Selenium: 4.0.0-alpha-6 

## Testcase

Trying to run a session on firefox android causes server error with geckodriver. The session doesn't get started

## Stacktrace

`1615816644018	webdriver::server	DEBUG	<- 500 Internal Server Error {"value":{"error":"unknown error","message":"TypeError: this.curBrowser is null","stacktrace":"GeckoDriver.prototype.newSession@chrome://marionette/content/driver.js:910:7\ndespatch@chrome://marionette/content/server.js:297:40\nexecute@chrome://marionette/content/server.js:267:16\nonPacket/<@chrome://marionette/content/server.js:240:20\nonPacket@chrome://marionette/content/server.js:241:9\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:504:20\n"}}`

## Trace-level log

```
1615816635402	geckodriver::android	DEBUG	Pushing GeckoView configuration file to /data/local/tmp/org.mozilla.firefox-geckoview-config.yaml
1615816635412	mozdevice	DEBUG	execute_host_command: >> "host:transport:RZ8M40MGAVB"
1615816635412	mozdevice	DEBUG	execute_host_command: << []
1615816635412	mozdevice	DEBUG	execute_host_command: >> "shell:chmod a+rw /data/local/tmp/org.mozilla.firefox-geckoview-config.yaml"
1615816635460	mozdevice	DEBUG	execute_host_command: << ""
1615816635461	mozdevice	DEBUG	execute_host_command: >> "host:transport:RZ8M40MGAVB"
1615816635461	mozdevice	DEBUG	execute_host_command: << []
1615816635461	mozdevice	DEBUG	execute_host_command: >> "shell:am set-debug-app --persistent org.mozilla.firefox"
1615816635546	mozdevice	DEBUG	execute_host_command: << ""
1615816635546	geckodriver::android	DEBUG	Launching org.mozilla.firefox/.App
1615816635548	mozdevice	DEBUG	execute_host_command: >> "host:transport:RZ8M40MGAVB"
1615816635548	mozdevice	DEBUG	execute_host_command: << []
1615816635548	mozdevice	DEBUG	execute_host_command: >> "shell:am start -W -n org.mozilla.firefox/.App --es args --marionette\\ --profile\\ /mnt/sdcard/org.mozilla.firefox-geckodriver-profile"
1615816637365	mozdevice	DEBUG	execute_host_command: << "Starting: Intent { cmp=org.mozilla.firefox/.App (has extras) }\nStatus: ok\nLaunchState: COLD\nActivity: org.mozilla.firefox/org.mozilla.fenix.HomeActivity\nTotalTime: 1704\nWaitTime: 1730\nComplete\n"
1615816637365	geckodriver::marionette	DEBUG	Waiting 60s to connect to browser on 127.0.0.1:63093
1615816643534	geckodriver::marionette	DEBUG	Connection to Marionette established on 127.0.0.1:63093.
1615816643565	webdriver::server	DEBUG	Deleting session
1615816643565	geckodriver::android	DEBUG	Force stopping the Android package: org.mozilla.firefox
1615816643565	mozdevice	DEBUG	Force stopping Android package: org.mozilla.firefox
1615816643566	mozdevice	DEBUG	execute_host_command: >> "host:transport:RZ8M40MGAVB"
1615816643566	mozdevice	DEBUG	execute_host_command: << []
1615816643566	mozdevice	DEBUG	execute_host_command: >> "shell:am force-stop org.mozilla.firefox"
1615816643827	mozdevice	DEBUG	execute_host_command: << ""
1615816643827	geckodriver::marionette	DEBUG	Android package force-stopped
1615816643829	mozdevice	DEBUG	execute_host_command: >> "host:transport:RZ8M40MGAVB"
1615816643829	mozdevice	DEBUG	execute_host_command: << []
1615816643829	mozdevice	DEBUG	execute_host_command: >> "shell:am clear-debug-app org.mozilla.firefox"
1615816643969	mozdevice	DEBUG	execute_host_command: << ""
1615816643969	geckodriver::android	DEBUG	Disabled reading from configuration file
1615816643971	mozdevice	DEBUG	execute_host_command: >> "host:transport:RZ8M40MGAVB"
1615816643971	mozdevice	DEBUG	execute_host_command: << []
1615816643971	mozdevice	DEBUG	execute_host_command: >> "shell:rm -rf /data/local/tmp/org.mozilla.firefox-geckoview-config.yaml"
1615816644014	mozdevice	DEBUG	execute_host_command: << ""
1615816644014	geckodriver::android	DEBUG	Deleted GeckoView configuration file
1615816644017	mozdevice	DEBUG	execute_host_command: >> "host:transport:RZ8M40MGAVB"
1615816644017	mozdevice	DEBUG	execute_host_command: << []
1615816644017	mozdevice	DEBUG	execute_host_command: >> "killforward:tcp:63093"
1615816644018	geckodriver::android	ERROR	Android port forward (63093 -> 2829) failed to stop: adb error: closed
1615816644018	webdriver::server	DEBUG	<- 500 Internal Server Error {"value":{"error":"unknown error","message":"TypeError: this.curBrowser is null","stacktrace":"GeckoDriver.prototype.newSession@chrome://marionette/content/driver.js:910:7\ndespatch@chrome://marionette/content/server.js:297:40\nexecute@chrome://marionette/content/server.js:267:16\nonPacket/<@chrome://marionette/content/server.js:240:20\nonPacket@chrome://marionette/content/server.js:241:9\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:504:20\n"}}

```

**Comment by whimboo** [comment]
@sparsh00 is that a permanent failure? I wonder if you could test with a Nightly build of Firefox, and if maybe my patch on https://bugzilla.mozilla.org/show_bug.cgi?id=1696878 fixed that. If that is the case I would request uplift for the patch.

Also is that new for you with Firefox 86?

**Comment by whimboo** [comment]
You could download the apk from here:
https://firefox-ci-tc.services.mozilla.com/tasks/index/mobile.v2.fenix.nightly.latest

**Comment by sparsh00** [comment]
@whimboo Yes is reproducible everytime. I only tried on 86.1.1 on my device.
I tried the nightly build but it crashes as soon as i start remote debugging via usb for Marionette to connect.

**Comment by whimboo** [comment]
That file should always be deleted by geckodriver when the test finished. So I wonder in which cases it remains. Oh, and I see that you are using geckodriver 0.27. Please update to 0.29 and try again. 

When you say Nightly is working then what does it mean? Did you start it manually, or does it really mean it works when it gets started by geckodriver?

Also please try the 85 release. I would really like to know if that one works:
https://github.com/mozilla-mobile/fenix/releases/tag/v85.1.0

Or does 86 also work after you deleted the above file?

**Comment by sparsh00** [comment]
Tried running with 0.29.0 with `--android-storage sdcard` (as without that firefox crashes, also my phone doesn't have sdcard. Might raise another issue for this) and firefox 85.1.0 and getting the same error.
For nightly i mean it was not even working manually when the file was present and doesn't work with geckodriver in any case. Doesn't even get `Connection to Marionette established`

**Comment by whimboo** [comment]
Ok, so I would say lets continue with Nightly so that we can get the various issues fixed and verified for you. Btw did you try the various other `--android-storage` options? Which Android version do you have?

It would be great if you could attach a [trace log](https://firefox-source-docs.mozilla.org/testing/geckodriver/TraceLogs.html) from geckodriver, and also the output from `adb` (please attach both as files and don't add as comment).

**Comment by sparsh00** [comment]
Sure will continue with nightly, here are some logs for different `--android-storage` options- 

[adb_nightly_app.log](https://github.com/mozilla/geckodriver/files/6147049/adb_nightly_app.log)
[geckodriver_nightly_app.log](https://github.com/mozilla/geckodriver/files/6147052/geckodriver_nightly_app.log)

[adb_nightly_internal.log](https://github.com/mozilla/geckodriver/files/6147054/adb_nightly_internal.log)
[geckodriver_noghtly_intermal.log](https://github.com/mozilla/geckodriver/files/6147055/geckodriver_noghtly_intermal.log)

[adb_nightly_sdcard.log](https://github.com/mozilla/geckodriver/files/6147056/adb_nightly_sdcard.log)
[geckodriver_nightly_sdcard.log](https://github.com/mozilla/geckodriver/files/6147057/geckodriver_nightly_sdcard.log)

And logs after I delete file manually `rm -rf /data/local/tmp/org.mozilla.fenix-geckoview-config.yaml` and start the apk manually from device

[nightly_after_delete_manually.log](https://github.com/mozilla/geckodriver/files/6147069/nightly_after_delete_manually.log)

Also can see from logs that `marionette` is not getting started.

**Comment by whimboo** [comment]
In all the cases above the `adb` log shows a crash during the early startup of Firefox. That also happens when not running manually. I wonder if that is also caused by the config file. Could you explicitly check how geckodriver works when this config file is not present before starting it? Does it also cause a crash of Firefox?

**Comment by whimboo** [comment]
Also for a better investigation feel free to join us on https://chat.mozilla.org/#/room/#interop:mozilla.org.

**Comment by sparsh00** [comment]
Sure will look into it. By how geckodriver works, do you mean geckodriver inside firefox apk or geckodriver running on my local machine which tries and start automate session. I am new to geckodriver and android dev so could you help me with where to look and mobile apk repo

**Comment by whimboo** [comment]
I meant that you should make sure to explicitly remove `/data/local/tmp/org.mozilla.fenix-geckoview-config.yaml` before starting your test with geckodriver. Also an information I still miss is the version of Android you are running with.

**Comment by sparsh00** [comment]
Yup have checked manually that its not present before running in the above tests. I am running on android v10. I have a samsung device with their one ui if it matters

**Comment by whimboo** [comment]
After some investigation I can reproduce the problem. I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1699065 to get this fixed in Marionette.

**Comment by Dharin-shah** [comment]
Please note as an FYI, this works on emulators and rooted devices, just not working on non-rooted real devices

**Comment by whimboo** [comment]
The problem here was not in Marionette but geckodriver itself. By default it was using the wrong activity. With the geckodriver 0.29.1 release today this should no longer be a problem. Just specify the package name, and geckodriver will pick defaults for the activity and the intent arguments.
